File: csharp/EmbeddedSpiceDB.cs
using System.Net.Http;
using System.Net.Sockets;
using System.Net;
using Authzed.Api.V1;
using Grpc.Core;
using Grpc.Net.Client;

namespace Rendil.Spicedb.Embedded;

/// <summary>
/// Embedded SpiceDB instance.
/// A thin wrapper that starts SpiceDB via the C-shared library, connects over a Unix
/// socket, and bootstraps schema and relationships via gRPC.
/// Use Permissions(), Schema(), and Watch() to access the full SpiceDB API.
/// </summary>
public sealed class EmbeddedSpiceDB : IDisposable
{
    private readonly ulong _handle;
    private readonly GrpcChannel _channel;
    private bool _disposed;

    private EmbeddedSpiceDB(ulong handle, GrpcChannel channel)
    {
        _handle = handle;
        _channel = channel;
    }

    /// <summary>
    /// Create a new embedded SpiceDB instance with a schema and relationships.
    /// </summary>
    /// <param name="schema">The SpiceDB schema definition (ZED language)</param>
    /// <param name="relationships">Initial relationships (empty list or null allowed)</param>
    /// <returns>New EmbeddedSpiceDB instance</returns>
    public static EmbeddedSpiceDB Create(string schema, IReadOnlyList<Relationship>? relationships = null)
    {
        var (handle, socketPath) = SpiceDBFFI.Start();

        var httpHandler = new SocketsHttpHandler
        {
            ConnectCallback = async (ctx, ct) =>
            {
                var socket = new Socket(AddressFamily.Unix, SocketType.Stream, ProtocolType.IP);
                var endpoint = new UnixDomainSocketEndPoint(socketPath);
                await socket.ConnectAsync(endpoint, ct);
                return new NetworkStream(socket, ownsSocket: true);
            },
        };

        var httpClient = new HttpClient(httpHandler) { BaseAddress = new Uri("http://localhost") };

        var options = new GrpcChannelOptions
        {
            HttpClient = httpClient,
            Credentials = ChannelCredentials.Insecure,
        };

        GrpcChannel channel;
        try
        {
            channel = GrpcChannel.ForAddress("http://localhost", options);
        }
        catch (Exception ex)
        {
            try { SpiceDBFFI.Dispose(handle); } catch { /* best effort */ }
            throw new SpiceDBException("Failed to connect to SpiceDB: " + ex.Message, ex);
        }

        var db = new EmbeddedSpiceDB(handle, channel);

        try
        {
            db.Bootstrap(schema, relationships ?? Array.Empty<Relationship>());
        }
        catch (Exception ex)
        {
            db.Dispose();
            throw new SpiceDBException("Failed to bootstrap: " + ex.Message, ex);
        }

        return db;
    }

    private void Bootstrap(string schema, IReadOnlyList<Relationship> relationships)
    {
        var schemaClient = new SchemaService.SchemaServiceClient(_channel);
        schemaClient.WriteSchema(new WriteSchemaRequest { Schema = schema });

        if (relationships.Count > 0)
        {
            var permClient = new PermissionsService.PermissionsServiceClient(_channel);
            var updates = relationships.Select(r => new RelationshipUpdate
            {
                Operation = RelationshipUpdate.Types.Operation.Touch,
                Relationship = r,
            }).ToList();
            permClient.WriteRelationships(new WriteRelationshipsRequest { Updates = { updates } });
        }
    }

    /// <summary>
    /// Permissions service client (CheckPermission, WriteRelationships, ReadRelationships, etc.).
    /// </summary>
    public PermissionsService.PermissionsServiceClient Permissions() =>
        new PermissionsService.PermissionsServiceClient(_channel);

    /// <summary>
    /// Schema service client (ReadSchema, WriteSchema, ReflectSchema, etc.).
    /// </summary>
    public SchemaService.SchemaServiceClient Schema() =>
        new SchemaService.SchemaServiceClient(_channel);

    /// <summary>
    /// Watch service client for relationship changes.
    /// </summary>
    public WatchService.WatchServiceClient Watch() =>
        new WatchService.WatchServiceClient(_channel);

    /// <summary>
    /// Underlying gRPC channel for custom usage.
    /// </summary>
    public GrpcChannel Channel => _channel;

    /// <inheritdoc />
    public void Dispose()
    {
        if (_disposed) return;
        SpiceDBFFI.Dispose(_handle);
        _channel.Dispose();
        _disposed = true;
        GC.SuppressFinalize(this);
    }
}

File: java/src/main/java/com/rendil/spicedb/embedded/EmbeddedSpiceDB.java
package com.rendil.spicedb.embedded;

import com.authzed.api.v1.PermissionsServiceGrpc;
import com.authzed.api.v1.Relationship;
import com.authzed.api.v1.RelationshipUpdate;
import com.authzed.api.v1.SchemaServiceGrpc;
import com.authzed.api.v1.WatchServiceGrpc;
import com.authzed.api.v1.WriteRelationshipsRequest;
import com.authzed.api.v1.WriteSchemaRequest;
import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.sun.jna.Pointer;
import io.grpc.ManagedChannel;
import java.util.List;

/**
 * Embedded SpiceDB instance.
 *
 * <p>A thin wrapper that starts SpiceDB via the C-shared library (shared/c), connects over a Unix
 * socket, and bootstraps schema and relationships via gRPC.
 *
 * <p>Use {@link #permissions()}, {@link #schema()}, and {@link #watch()} to access the full SpiceDB
 * API.
 *
 * <p>Prerequisites: Build shared/c first ({@code mise run shared-c-build}).
 */
public final class EmbeddedSpiceDB implements AutoCloseable {

  private final long handle;
  private final ManagedChannel channel;

  private EmbeddedSpiceDB(long handle, ManagedChannel channel) {
    this.handle = handle;
    this.channel = channel;
  }

  /**
   * Create a new embedded SpiceDB instance with a schema and relationships.
   *
   * @param schema The SpiceDB schema definition (ZED language)
   * @param relationships Initial relationships (empty list allowed)
   * @return New EmbeddedSpiceDB instance
   */
  public static EmbeddedSpiceDB create(String schema, List<Relationship> relationships) {
    SpiceDB lib = SpiceDB.load();
    Pointer result = lib.spicedb_start();
    if (result == null) {
      throw new SpiceDBException("Null response from C library");
    }
    String json = result.getString(0);
    lib.spicedb_free(result);

    JsonObject parsed = new Gson().fromJson(json, JsonObject.class);
    if (parsed == null || !parsed.has("success") || !parsed.get("success").getAsBoolean()) {
      String err =
          parsed != null && parsed.has("error")
              ? parsed.getAsJsonPrimitive("error").getAsString()
              : "Unknown error";
      throw new SpiceDBException("Failed to start SpiceDB: " + err);
    }

    JsonObject data = parsed.getAsJsonObject("data");
    long handle = data.getAsJsonPrimitive("handle").getAsLong();
    String socketPath = data.getAsJsonPrimitive("socket_path").getAsString();

    ManagedChannel channel;
    try {
      channel = UnixSocketChannel.build(socketPath);
    } catch (Exception e) {
      // Dispose the instance on connection failure
      Pointer disposeResult = lib.spicedb_dispose(handle);
      if (disposeResult != null) {
        lib.spicedb_free(disposeResult);
      }
      throw new SpiceDBException("Failed to connect to SpiceDB: " + e.getMessage(), e);
    }

    EmbeddedSpiceDB db = new EmbeddedSpiceDB(handle, channel);

    // Bootstrap via gRPC
    var schemaStub = SchemaServiceGrpc.newBlockingStub(channel);
    schemaStub.writeSchema(WriteSchemaRequest.newBuilder().setSchema(schema).build());

    if (relationships != null && !relationships.isEmpty()) {
      var permStub = PermissionsServiceGrpc.newBlockingStub(channel);
      var updates =
          relationships.stream()
              .map(
                  r ->
                      RelationshipUpdate.newBuilder()
                          .setOperation(RelationshipUpdate.Operation.OPERATION_TOUCH)
                          .setRelationship(r)
                          .build())
              .toList();
      permStub.writeRelationships(
          WriteRelationshipsRequest.newBuilder().addAllUpdates(updates).build());
    }

    return db;
  }

  /**
   * @return Permissions service stub (CheckPermission, WriteRelationships, ReadRelationships, etc.)
   */
  public PermissionsServiceGrpc.PermissionsServiceBlockingStub permissions() {
    return PermissionsServiceGrpc.newBlockingStub(channel);
  }

  /**
   * @return Schema service stub (ReadSchema, WriteSchema, ReflectSchema, etc.)
   */
  public SchemaServiceGrpc.SchemaServiceBlockingStub schema() {
    return SchemaServiceGrpc.newBlockingStub(channel);
  }

  /**
   * @return Watch service stub (watch for relationship changes)
   */
  public WatchServiceGrpc.WatchServiceBlockingStub watch() {
    return WatchServiceGrpc.newBlockingStub(channel);
  }

  /**
   * @return The underlying gRPC channel for custom usage
   */
  public ManagedChannel channel() {
    return channel;
  }

  @Override
  public void close() {
    SpiceDB lib = SpiceDB.load();
    Pointer result = lib.spicedb_dispose(handle);
    if (result != null) {
      lib.spicedb_free(result);
    }
    channel.shutdown();
  }
}
